### PR TITLE
Close #20 Sign Up

### PR DESCRIPTION
**Pull request by HaiTranVar**

### CODE REVIEW

File: app/views/password_resets/edit.html.erb [hunk 1]
```diff
@@ -6,7 +6,7 @@
       <%= hidden_field_tag :email, @user.email %>
 
       <div class="form-group">
-        <%= f.label :password %>
+        <%= f.label :password, "Password" %>
```
Comment (HaiTranVar) [inline app/views/password_resets/edit.html.erb]: i18n

File: app/controllers/account_activations_controller.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+class AccountActivationsController < ApplicationController
+
+  def edit
+    email_param = {email: params[:email]}
+    user = User.find_by email_param
+
```
Comment (HaiTranVar) [inline app/controllers/account_activations_controller.rb]: remove this line

File: app/controllers/account_activations_controller.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    if user && !user.activated? && user.authenticated?(:activation, params[:id])
+      user.update_attribute(:activated,    true)
+      user.update_attribute(:activated_at, Time.zone.now)
+      log_in user
+      flash[:success] = "Account activated!"
```
Comment (HaiTranVar) [inline app/controllers/account_activations_controller.rb]: i18n

File: app/controllers/account_activations_controller.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+      redirect_to login_path
+    else
+      flash[:danger] = "Invalid activation link"
```
Comment (HaiTranVar) [inline app/controllers/account_activations_controller.rb]: i18n

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -3,14 +3,20 @@ def new
...
-      log_in @user
-      flash[:success] = t ".success_login"
-      redirect_to root_path
+    user = User.find_by(email: params[:session][:email].downcase)
+    if user && user.authenticate(params[:session][:password])
+      if user.activated?
+        log_in user
+        redirect_to root_path
+      else
+        message  = "Account not activated. "
```
Comment (HaiTranVar) [inline app/controllers/sessions_controller.rb]: i18n

File: app/controllers/sessions_controller.rb [hunk 6]
```diff
@@ -3,14 +3,20 @@ def new
...
+        message += "Check your email for the activation link."
```
Comment (HaiTranVar) [inline app/controllers/sessions_controller.rb]: i18n

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -3,14 +3,20 @@ def new
...
+        flash[:warning] = message
+        redirect_to root_url
+      end
     else
-      flash.now[:danger] = t ".error_login"
-      render "new"
+      flash.now[:danger] = 'Invalid email/password combination'
+      render 'new'
```
Comment (HaiTranVar) [inline app/controllers/sessions_controller.rb]: ' > ''

File: app/controllers/sessions_controller.rb [hunk 8]
```diff
@@ -3,14 +3,20 @@ def new
...
+      else
+        message  = "Account not activated. "
+        message += "Check your email for the activation link."
+        flash[:warning] = message
+        redirect_to root_url
+      end
     else
-      flash.now[:danger] = t ".error_login"
-      render "new"
+      flash.now[:danger] = 'Invalid email/password combination'
```
Comment (HaiTranVar) [inline app/controllers/sessions_controller.rb]: i18n

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -1,13 +1,13 @@
 class UsersController < ApplicationController
-  before_action :check_is_logged_in, :check_is_admin_permission
+  #before_action :check_is_logged_in, :check_is_admin_permission
   before_action :get_user, only: [:edit, :update, :destroy, :show]
 
   def index
-    @users = User.paginate page: params[:page]
+    @users = User.paginate(page: params[:page])
```
Comment (HaiTranVar) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -17,8 +17,9 @@ def new
   def create
     @user = User.new user_params
     if @user.save
-      flash[:success] = t ".success_create"
-      redirect_to @user
+      UserMailer.account_activation(@user).deliver_now
+      flash[:info] = "Please check your email to activate your account."
```
Comment (HaiTranVar) [inline app/controllers/users_controller.rb]: i18n

File: app/helpers/account_activations_helper.rb [hunk 11]
```diff
@@ -0,0 +1,2 @@
+module AccountActivationsHelper
+end
```
Comment (HaiTranVar) [inline app/helpers/account_activations_helper.rb]: remove this file

File: app/mailers/user_mailer.rb [hunk 12]
```diff
@@ -1,10 +1,13 @@
 class UserMailer < ApplicationMailer
-
   # Subject can be set in your I18n file at config/locales/en.yml
   # with the following lookup:
   #
   #   en.user_mailer.password_reset.subject
   #
+  def account_activation(user)
+    @user = user
+    mail to: @user.email, subject: "Account activation"
```
Comment (HaiTranVar) [inline app/mailers/user_mailer.rb]: i18n

File: app/models/user.rb [hunk 13]
```diff
@@ -19,9 +22,16 @@ def create_reset_digest
...
-  def authenticated? token
-    return if reset_digest.nil?
-    BCrypt::Password.new(reset_digest).is_password?(token)
+  # Returns true if the given token matches the digest.
+  def authenticated?(attribute, token)
+    digest = send("#{attribute}_digest")
+    return false if digest.nil?
+    BCrypt::Password.new(digest).is_password?(token)
+  end
+  # Remembers a user in the database for use in persistent sessions.
```
Comment (HaiTranVar) [inline app/models/user.rb]: remove this function below

File: app/models/user.rb [hunk 14]
```diff
@@ -19,9 +22,16 @@ def create_reset_digest
     update_attribute :reset_send_at, Time.zone.now
   end
 
-  def authenticated? token
-    return if reset_digest.nil?
-    BCrypt::Password.new(reset_digest).is_password?(token)
+  # Returns true if the given token matches the digest.
```
Comment (HaiTranVar) [inline app/models/user.rb]: romove the comment

File: app/models/user.rb [hunk 15]
```diff
@@ -39,4 +49,25 @@ def new_token
       SecureRandom.urlsafe_base64
     end
   end
+  # Activates an account.
```
Comment (HaiTranVar) [inline app/models/user.rb]: remove 2 functions below

File: app/views/password_resets/edit.html.erb [hunk 16]
```diff
@@ -6,12 +6,12 @@
...
 
       <div class="form-group">
-        <%= f.label :password, t(".password") %>
+        <%= f.label :password, "Password" %>
         <%= f.password_field :password, class: "form-control", require: true %>
       </div>
 
       <div class="form-group">
-        <%= f.label :password_confirm, t(".password_comfirm") %>
+        <%= f.label :password_confirm %>
```
Comment (HaiTranVar) [inline app/views/password_resets/edit.html.erb]: add password confirm

File: app/views/password_resets/new.html.erb [hunk 17]
```diff
@@ -4,7 +4,7 @@
   <div class="col-md-4 col-md-offset-4">
     <%= form_for(:password_reset, url: password_resets_path) do |f| %>
       <div class="form-group">
-        <%= f.label :email, t(".email") %>
+        <%= f.label :email %>
```
Comment (HaiTranVar) [inline app/views/password_resets/new.html.erb]: add email

File: app/views/sessions/new.html.erb [hunk 18]
```diff
@@ -15,6 +15,7 @@
 
       <div class="form-group">
         <%= link_to t(".forget_password"), new_password_reset_path %>
+        <%= link_to "Sign up now!", new_user_path %>
```
Comment (HaiTranVar) [inline app/views/sessions/new.html.erb]: i18n

File: app/views/user_mailer/account_activation.html.erb [hunk 19]
```diff
@@ -0,0 +1,8 @@
+<h1>Online Exam</h1>
```
Comment (HaiTranVar) [inline app/views/user_mailer/account_activation.html.erb]: i18n

File: app/views/user_mailer/account_activation.text.erb [hunk 20]
```diff
@@ -0,0 +1,3 @@
+Hi <%= @user.name %>,
```
Comment (HaiTranVar) [inline app/views/user_mailer/account_activation.text.erb]: i18n

File: app/views/users/new.html.erb [hunk 21]
```diff
@@ -1,7 +1,24 @@
...
+<% provide(:title, 'Sign up') %>
+<h1 class="text-center">Sign up</h1>
+
 <div class="row">
   <div class="col-md-6 col-md-offset-3">
-    <%= render "form", user: @user %>
+    <%= form_for(@user) do |f| %>
+      <%= render 'shared/error_messages_signup' %>
+
+      <%= f.label :name %>
```
Comment (HaiTranVar) [inline app/views/users/new.html.erb]: add a name

File: app/controllers/account_activations_controller.rb [hunk 22]
```diff
@@ -1,16 +1,22 @@
...
     else
-      flash[:danger] = t ".invalid_activation_link"
+      flash[:danger] = t '.invalid_activation_link'
       redirect_to login_path
     end
   end
+
+  private
+
+  def active_account
```
Comment (dinhngochien) [inline app/controllers/account_activations_controller.rb]: - recommend rename to: `user_active_account` for clear
- the method needs a parameter
- it's better to place the method in `user.rb` model

File: app/controllers/account_activations_controller.rb [hunk 23]
```diff
@@ -1,16 +1,22 @@
...
+  before_action :active_account, only: %i[edit]
   def edit
     user = User.find_by(email: params[:email])
     if user && !user.activated? && user.authenticated?(:activation, params[:id])
-      user.update_attribute(:activated,    true)
-      user.update_attribute(:activated_at, Time.zone.now)
-      log_in user
-      flash[:success] = t ".account_activated"
-      redirect_to login_path
+      active_account
```
Comment (dinhngochien) [inline app/controllers/account_activations_controller.rb]: pass `user` to this method

File: app/controllers/sessions_controller.rb [hunk 24]
```diff
@@ -6,19 +6,11 @@ def new
...
-        log_in user
-        redirect_to root_path
-      else
-        message  = t ".account_not_activated"
-        message += t ".check_your_email"
-        flash[:warning] = message
-        redirect_to root_url
-      end
+    if user&.authenticate(params[:session][:password])
+      user_active
```
Comment (dinhngochien) [inline app/controllers/sessions_controller.rb]: pass `user` to this method

File: app/controllers/sessions_controller.rb [hunk 25]
```diff
@@ -32,4 +24,16 @@ def destroy
   def get_user
     @user = User.find_by email: params[:session][:email]
   end
+
+  def user_active
```
Comment (dinhngochien) [inline app/controllers/sessions_controller.rb]: - rename to `is_activated?`
- the method needs a parameter
- you can place this to `user.rb` model also

File: app/controllers/users_controller.rb [hunk 26]
```diff
@@ -64,4 +57,15 @@ def get_user
     flash[:danger] = t 'error_404'
     redirect_to root_path
   end
+
+  def user_save
```
Comment (dinhngochien) [inline app/controllers/users_controller.rb]: the method name does not make sense, please update to `activate_account`

**Pull request merged**